crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder/automatic_shield_builder.rs

```diff
@@ -0,0 +1,916 @@
+        self.proto_matrix.factors_for_role(role)
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_category(
+        &mut self,
+        to: RoleKind,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        match self.assign_factors_matching_selector(
+            to,
+            FactorSelector::Category(category),
+            quantity_to_add,
+        ) {
+            Some(0) | None => false,
+            Some(_) => true,
+        }
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_kind(
+        &mut self,
+        to: RoleKind,
+        kind: FactorSourceKind,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        match self.assign_factors_matching_selector(
+            to,
+            FactorSelector::Kind(kind),
+            quantity_to_add,
+        ) {
+            Some(0) | None => false,
+            Some(_) => true,
+        }
+    }
+
+    fn count_factors_for_role(&self, role_kind: RoleKind) -> u8 {
+        self.factors_for_role(role_kind).len() as u8
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_category_to_recovery(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        self.assign_factors_of_category(Recovery, category, quantity_to_add)
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_category_to_confirmation(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        self.assign_factors_of_category(Confirmation, category, quantity_to_add)
+    }
+
+    fn assign_factor_of_category_to_role_while_meaningful_and_less_than_limit(
+        &mut self,
+        category: FactorSourceCategory,
+        limit: u8,
+        to: RoleKind,
+    ) {
+        let role = to;
+
+        loop {
+            if self.count_factors_for_role(role) >= limit {
+                // when `limit` reached, we stop.
+                return;
+            }
+
+            if !category.is_supported() {
+                self.stats_for_testing
+                    .calls_to_assign_unsupported_factor
+                    .push(CallsToAssignUnsupportedFactor {
+                        role,
+                        unsupported: FactorSelector::Category(category),
+                        number_of_factors_for_role: self
+                            .count_factors_for_role(role),
+                        limit,
+                    });
+                return;
+            }
+
+            if !self.assign_factors_of_category(role, category, Quantity::One) {
+                // We did not manage to assign any hardware factor, meaning we
+                // it is meaningless to try to assign more factors of this category.
```

Comment:
Why hardware? Cannot infer this from the code as the category parameter can be any category.